backend/apps/events/api/v1/views.py

```diff
@@ -0,0 +1,118 @@
+from apps.events.api.v1.filters import EventFilter
+from apps.events.api.v1.paginators import CustomEventPagination
+from apps.events.api.v1.serializers import EventListSerializer, EventRUDSerializer
+from apps.events.models import Event
+from django.db.models import Case, Count, When
+from django.db.utils import IntegrityError
+from django.utils import timezone
+from django_filters import rest_framework as filters
+from rest_framework import status
+from rest_framework.generics import GenericAPIView
+from rest_framework.mixins import (
+    CreateModelMixin,
+    DestroyModelMixin,
+    ListModelMixin,
+    RetrieveModelMixin,
+    UpdateModelMixin,
+)
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.viewsets import GenericViewSet
+
+
+class ListCreateEventAPIView(ListModelMixin, CreateModelMixin, GenericViewSet):
+    """
+    LIST / CREATE view for the Event model.
+
+    API
+    -----------
+    get:
+    Return a list of Events.
+
+    post:
+    Create a new Event instance.
+
+    """
+
+    serializer_class = EventListSerializer
+    pagination_class = CustomEventPagination
+    permission_classes = (IsAuthenticated,)
+    filter_backends = (filters.DjangoFilterBackend,)
+    filterset_class = EventFilter
+
+    def get_queryset(self):
+        current_time = timezone.now()
+        return (
+            Event.objects.all()
+            .order_by("start_time")
+            .annotate(
+                visitors=Count("staff_participants") + Count("candidate_participants"),
+                is_future=Case(When(start_time__gt=current_time, then=True), default=False),
+            )
+        )
+
+    def create(self, request, *args, **kwargs):
+        """Check that none of the db constraints was violated."""
+        try:
+            serializer = self.get_serializer(data=request.data)
+            serializer.is_valid(raise_exception=True)
+            self.perform_create(serializer)
+            headers = self.get_success_headers(serializer.data)
+            return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+        except IntegrityError:
+            return Response(
+                data={
+                    "message": "Can't create event if it's end time is greater than "
+                    "it's start time."
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
```

Comment:
I think, we must check these dates inside the serializer's method `validate()` and also throw the ValidationError there if needed.

Reply:
I put a constraint on the db, so that it never allows to put wrong data here. So now it works in Admin site, Factories, shell, etc. Here I specifically catch it not to crash the app, and so there is no need to check it in validate() one more time. Still, it is a good idea to catch validation errors as you are suggesting, I was under the impression that it does so under the hood. Now, I would need to check this later, thank you.

Reply:
I have just checked it, Serializer actually validates all the fields and there is no need to catch them.
So all is good =)

![image](https://user-images.githubusercontent.com/20648373/189398665-33dd488e-a576-4f22-99c4-394a62ddd151.png)